test/spec/php_spec.rb

```diff
@@ -0,0 +1,201 @@
+require_relative "spec_helper"
+
+describe "A PHP application" do
+	def self.genmatrix(matrix, keys = nil)
+		product_hash(matrix.select {|k,v| !keys || keys.include?(k) })
+	end
+	
+	def self.gencmd(args)
+		args.compact.map { |k,v|
+			if k.is_a? Numeric
+				ret = v.shellescape # it's an argument, so we want the value only
+			else
+				ret = k.shellescape
+				unless !!v == v # check if boolean
+					ret.concat(" #{v.shellescape}") # --foobar flags have no values
+				end
+			end
+			ret
+		}.join(" ").strip
+	end
+	
+	# the matrix of options and arguments to test
+	# we will generate all possible combinations of these using a helper
```

Comment:
Can you help me understand why running through every combination here is valuable?

`"conf/apache2.server.include.broken"` is going to be run 12 times. Does that provide more certainty than if it was run once or twice?

48 deploys for apache and 36 for nginx is a lot. Adding more examples of broken config will make this grow geometrically, and it becomes more likely that some spurious error will cause your tests to fail. Can the same testing be accomplished in a handful of runs?

Edit: Or are all of these being executed with `app.run` once an app is already deployed?

Reply:
Sorry, the comment is incorrect. We're not generating all possible combinations, but just a smaller subset, further down, of the combinations that are interesting because they potentially interact.

And yes, the app for this is only deployed once.

Reply:
Do you need the fancy combinatorial functionality? Listing out the interesting combinations in separate tests would be simple and easier for others to follow. I like to keep things simple whenever possible :)